### ISSUE TITLE
Closure captures future variable resulting in <internal structure>

### ISSUE THREAD

**Issue by dtcaciuc** [elm-lang/elm-compiler#1499]
This can be reproduced on Elm 0.17.1

The following problem came up because of a typo I made in the code. The snippet simply shuffles numbers from one list to another. In the function `f'`, I mistyped `yy` instead of `y` creating a cycle and everything still compiled. The program prints out `<internal structure>`. Should there be a guard against cases like these?

```
import Html exposing (text)

f =
  let
    f' : List number -> List number -> List number
    f' elems y =
      case elems of
        [] ->
          yy
        x :: rest ->
          f' rest (x :: yy)
          
    yy = f' [1, 2, 3] []
    
  in
    yy

main =
  text <| "Hello, World!" ++ (toString f)
```

**Comment by dtcaciuc** [comment]
Folks on elm-dev suggested my issue is a duplicate of #873

**Comment by jvoigtlaender** [comment]
I don't think it's a duplicate of #873, because one of the entities involved in your cycle is a function, not a non-function value. At least, the fix to #873 in https://github.com/elm-lang/elm-compiler/commit/8b484abfb313f2877403b168e4d7229863f20ba1 (according to the description in the commit comment) will not reject the program above at compile time.

I think the program is just buggy in a way that the compiler cannot catch. It's an issue for the programmer, not for the compiler repo.

**Comment by evancz** [comment]
I'm gonna trust @jvoigtlaender on this because I don't get your example. (Can you use reasonable variable names?!)

It is possible to write programs that diverge and programs that are generally weird in a Turing complete language and Rice's Theorem means we cannot catch them all. If you think this is something that can be caught, can you get it down to an http://sscce.org in a new issue with better names in your example code?
